ADRs/2022-12-08-PropertyWrappers-ADR.md

```diff
@@ -0,0 +1,184 @@
+            ...
+        }
+    }
+}
+```
+
+These are the problems with the current approach:
+
+1. Wiring up the subscription, including toting a subscription property, introduces repetitive boilerplate code to each VSM view.
+1. It's easy to incorrectly configure the state subscription.
+1. Accidentally capturing `self` within the `sink` closure will cause memory leaks (strong reference cycles)
+1. Using `state` or `container.state` within the `sink` closure of the `$state` publisher will give you the previous value, not the current value. (Due to the inherent behavior of the `@Published` state property in the `StateContainer`)
+
+## Proposed Solution
+
+### AutoRendered Property Wrapper
+
+The most appropriate solution is to create a property wrapper that encapsulates the state observation logic.
+
+This property wrapper, called `@AutoRendered` can access the enclosing UIView (or UIViewController) and invoke a new `render()` function on `ViewStateRendering`. The `render()` function is not needed or used in SwiftUI views.
+
+Example Usage
+
+```swift
+class SomeViewController: UIViewController, ViewStateRendering {
+    @AutoRendered var container: StateContainer<SomeViewState>
+    ...
+    func render() {
+        switch state {
+            ...
+        }
+    }
+}
+```
+
+The solution utilizes the same technique that powers SwiftUI's `@Published` property behavior, where the property wrapper parent can be accessed in a static subscript to add observation behavior to the wrapped value.
+
+Benefits
+
+- Removes 9 lines of boilerplate code (Solves problem #1)
+- Removes potential for incorrect state subscription (Solves problem #2)
+- Removes potential for memory leaks in state observation (Solves problem #3)
+- Invokes `render()` after the state value changes, ensuring `state` and `container.state` are both safe to access in the rendering code. (Solves problem #4)
+- Allows custom state subscription to be possible if required (by opting out of the `@AutoRendered` property)
+
+Drawbacks
+
+- Forgetting to apply `@AutoRendered` to the `container` property will result in runtime behavior issues (Property wrapper usage cannot be enforced by the compiler, as is the standard for all property wrappers like `@State`, `@ObservedObject`, etc.)
+
+## Alternatives Considered
+
+Two other solutions were considered which provide similar benefits while making the VSM framework more ergonomic.
+
+### ViewState Property Wrapper
+
+This solution focuses on addressing the problems in question while improving the ergonomics of the SwiftUI VSM view code. It does so by introducing a `@ViewState` property wrapper which encapsulates the `StateContainer` entirely from the engineer. It also provides the same convenient auto-observing behavior as `@AutoRender`, allowing it to be used for both SwiftUI and UIKit.
+
+All `StateContainer` properties and functions would then be accessed directly from the `@ViewState` wrapper by way of the `_` underscore prefix.
```

Comment:
Could the `StateContainer` be exposed directly as a projected value on the `ViewState` property wrapper instead of having to expose all of the operations through `ViewState` and relying on `_`?  This would fall in line with how people are already used to working with `SwiftUI` property wrappers.

Reply:
I think you're onto something here. My mind was set on forwarding the function calls to the `StateContainer` as the `ViewStateRendering` protocol does, but dynamic-member lookup to the state container might be more straightforward, changing the spelling to: `state.container.observe(...)` 🤔 

Let's discuss this further offline with @simonpierreroy .